### DISCUSSION TITLE
niceGUI: docker image for ARMv7 Processor (raspberry 3b)

### DISCUSSION THREAD

**Discussion by J0ergL** [zauberzeug/nicegui#3432] [Q&A] [answered]
### Question

1st of all, niceGUI is in my opinion awesome! 
Thank you for all the work to put into this!

To my _question_: 
I am trying to use niceGUI within a docker container on a raspberry but:
`docker pull zauberzeug/nicegui` 
returns:
`no matching manifest for linux/arm/v7 in the manifest list entries`

So I  tried the development.dockerfile as a base, inspired by #2929 but the build fails with:

```
` => ERROR [3/6] RUN curl -sSL https://install.python-poetry.org | python3 - &&     cd /usr/local/bin &&     ln -s ~/.local/bin/poetry &&     poetry conf  142.8s
------
 > [3/6] RUN curl -sSL https://install.python-poetry.org | python3 - &&     cd /usr/local/bin &&     ln -s ~/.local/bin/poetry &&     poetry config virtualenvs.create false:
141.2 Retrieving Poetry metadata
141.2
141.2 # Welcome to Poetry!`
141.2 Installing Poetry (1.8.3): Installing Poetry
141.2 Installing Poetry (1.8.3): An error occurred. Removing partial environment.
141.2 Poetry installation failed.
141.2 See /poetry-installer-error-wfs14l9p.log for error logs.
------

 1 warning found (use docker --debug to expand):
 - JSONArgsRecommended: JSON arguments recommended for CMD to prevent unintended behavior related to OS signals (line 17)
dockerfile:7

```

Now I am stuck trying to get niceGUI installed within a python slim image. I tried 3.8, 3.11 and 3.12.
All attempts failed at: `Building wheel for uvloop (pyproject.toml)`
with `_Warning: Package 'uvloop.handles' is absent from the `packages` configuration`
I am aware this is not caused by niceGUI, still I am hopping someone here can point me to instructions on how to solve this?

Thank you

**Comment by J0ergL** [comment]
I am still struggeling, anyone here with suggestions on how get NiceGUI running on arm v7

**Comment by falkoschindler** [comment]
Hi @J0ergL,
Currently we don't have an ARMv7 at hand. (In most projects we switched from Raspberry Pi to NVIDIA Jetson...) But we also think the problem is unrelated to NiceGUI.
Anyway, maybe someone from the community can help?

**Comment by rodja** [comment]
With this line you can test if poetry installation works on ARMv7:

```bash
docker run --rm -it python:3.8-slim bash -c "apt update && apt install curl build-essential -y && curl -sSL https://install.python-poetry.org | python3 -"
```

**Reply by J0ergL** [reply]
Hi @rodja,
Thanks for the hint, poetry installation fails. I tried it with both, 3.8 and 3.12. In both cases:
```
Installing Poetry (1.8.3): An error occurred. Removing partial environment.
Poetry installation failed.
See /poetry-installer-error-pheq4qi3.log for error logs.
```
The log is attached and contains a.o.:
```
× Building wheel for cffi (pyproject.toml) did not run successfully.
        │ exit code: 1
        ╰─> [39 lines of output]
```

I supposed this is where I have to do some digging...?

[poetry-installer-error-pheq4qi3.log](https://github.com/user-attachments/files/16669411/poetry-installer-error-pheq4qi3.log)

**Comment by J0ergL** [comment]
I decided to stop digging and use the ARM64 image.
I want to install other applications in that image.
I see it is debian based, but as appuser I am not able to install applications like vim etc. with apt. sudo is not install either.
How do you install for example vim in that image?
Thank you

**Reply by rodja** [reply, marked as answer]
The Docker container only switches to appuser in the entrypoint. You could replace the entrypoint, but thats not very secure. It's most recommended to just write your own Dockerfile to extend the base image. Something like this:

```Dockerfile
FROM zauberzeug/nicegui:2.1.0

RUN apt-get update && \
    DEBIAN_FRONTEND=noninteractive TZ=Etc/UTC \
    apt-get install -y --no-install-recommends \
    vim \
    && rm -rf /var/lib/apt/lists/* \
    && apt-get clean
```

Build with

```bash
docker build -t nicegui-vim . 
```

And run bash in the container with

```bash
docker run --rm -it nicegui-vim bash
```

**Reply by J0ergL** [reply]
Thank you for the perfect answer. It made me understand a lot more.
